swift/Sources/FlatBuffers/Verifier.swift

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright 2021 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Foundation
+// TODO: - Add docs
+
+///
+public struct Verifier {
+
+  fileprivate let _checkAlignment: Bool
+  fileprivate var _capacity: Int
+  fileprivate var _apparentSize: UOffset = 0
+  fileprivate var _tableCount = 0
+  fileprivate var _upperBound: Int = 0
+
+  internal var capacity: Int { _capacity }
+  internal var _depth = 0
+  internal var _buffer: ByteBuffer
+  internal let _options: VeriferOptions
+
+  ///
+  /// - Parameters:
+  ///   - buffer:
+  ///   - maxDepth:
+  ///   - maxTableCount:
+  ///   - maxApparentSize:
+  ///   - checkAlignment:
+  /// - Throws:
+  /// - Note: THIS IS AN EXPERIMENTAL FEATURE, USE ONLY WITH `sizedBuffers`
+  public init(
+    buffer: inout ByteBuffer,
+    options: VeriferOptions = .init(),
+    checkAlignment: Bool = true) throws
+  {
+    // TODO: - Maybe assert instead of throwing?
+    guard buffer.capacity < FlatBufferMaxSize else {
+      throw FlatbuffersErrors.exceedsMaxSizeAllowed
+    }
+
+    _buffer = buffer
+    _capacity = buffer.capacity
+    _checkAlignment = checkAlignment
+    _options = options
+  }
+
+  ///
+  public mutating func reset() {
+    _depth = 0
+    _tableCount = 0
+    _upperBound = 0
+  }
+
+  ///
+  /// - Parameters:
+  ///   - position:
+  ///   - type:
+  /// - Returns:
+  public mutating func isAligned<T>(position: Int, type: T.Type) throws {
+
+    /// If check alignment is false this mutating function doesnt continue
+    if !_checkAlignment { return }
+
+    /// advance pointer to position X
+    let ptr = _buffer._storage.memory.advanced(by: position)
+    /// Check if the pointer is aligned
+    if Int(bitPattern: ptr) & (MemoryLayout<T>.alignment &- 1) == 0 {
+      return
+    }
+    throw FlatbuffersErrors.missAlignedPointer(
+      position: position,
+      type: String(describing: T.self))
+  }
+
+  ///
+  /// - Parameters:
+  ///   - position:
+  ///   - size:
+  /// - Returns:
+  public mutating func rangeInBuffer(position: Int, size: Int) throws {
+    let end = position &+ size
+    #warning("Check if this is correct")
+    if end < 0 || end > _buffer.capacity {
+      throw FlatbuffersErrors.outOfBounds(position: position, end: end)
+    }
```

Comment:
I am sure if the end should be less than zero, but I would assume so?

Reply:
Is `end` signed here, and could it be less than zero due to overflow?

Reply:
I'm checking now, altough i usually get a -2 in the fuzzer so i am not really sure.
In general the tests work, and the fuzzer passes

Reply:
I'll look into this a bit more though

Reply:
It seems that it is overflowing due to the function we have below,  https://github.com/google/flatbuffers/pull/6373#discussion_r615296516. where the value sometimes does become negative.

Reply:
Yes, end is a singed Int

Reply:
After looking into the code and checking the rust code again, and running the fuzzer again. it seems that end is indeed overflowing. and that is mainly because of the function `derefOffset` that adds an offset and reports an overflow. Unlike rust, we dont have a saturated add in swift. 

> Saturating integer subtraction. Computes self - rhs, saturating at the numeric bounds instead of overflowing.